File: paimon-core/src/main/java/org/apache/paimon/table/RollbackHelper.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.table;

import org.apache.paimon.Changelog;
import org.apache.paimon.Snapshot;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.utils.ChangelogManager;
import org.apache.paimon.utils.SnapshotManager;
import org.apache.paimon.utils.TagManager;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.SortedMap;

import static org.apache.paimon.utils.Preconditions.checkNotNull;

/** Helper class for {@link Table#rollbackTo} including utils to clean snapshots. */
public class RollbackHelper {

    private final SnapshotManager snapshotManager;
    private final ChangelogManager changelogManager;
    private final TagManager tagManager;
    private final FileIO fileIO;

    public RollbackHelper(
            SnapshotManager snapshotManager,
            ChangelogManager changelogManager,
            TagManager tagManager,
            FileIO fileIO) {
        this.snapshotManager = snapshotManager;
        this.changelogManager = changelogManager;
        this.tagManager = tagManager;
        this.fileIO = fileIO;
    }

    /** Clean snapshots and tags whose id is larger than given snapshot's and update latest hit. */
    public void cleanLargerThan(Snapshot retainedSnapshot) {
        try {
            cleanSnapshots(retainedSnapshot);
            cleanLongLivedChangelogs(retainedSnapshot);
            cleanTags(retainedSnapshot);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void createSnapshotFileIfNeeded(Snapshot taggedSnapshot) {
        // it is possible that the earliest snapshot is later than the rollback tag because of
        // snapshot expiration, in this case the `cleanLargerThan` method will delete all
        // snapshots, so we should write the tag file to snapshot directory and modify the
        // earliest hint
        if (!snapshotManager.snapshotExists(taggedSnapshot.id())) {
            try {
                fileIO.writeFile(
                        snapshotManager.snapshotPath(taggedSnapshot.id()),
                        taggedSnapshot.toJson(),
                        false);
                snapshotManager.commitEarliestHint(taggedSnapshot.id());
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
    }

    private void cleanSnapshots(Snapshot retainedSnapshot) throws IOException {
        long earliest =
                checkNotNull(
                        snapshotManager.earliestSnapshotId(), "Cannot find earliest snapshot.");
        long latest =
                checkNotNull(snapshotManager.latestSnapshotId(), "Cannot find latest snapshot.");

        // modify the latest hint
        snapshotManager.commitLatestHint(retainedSnapshot.id());

        // it is possible that some snapshots have been expired
        long to = Math.max(earliest, retainedSnapshot.id() + 1);
        for (long i = latest; i >= to; i--) {
            // Ignore the non-existent snapshots
            if (snapshotManager.snapshotExists(i)) {
                snapshotManager.deleteSnapshot(i);
            }
        }
    }

    private void cleanLongLivedChangelogs(Snapshot retainedSnapshot) throws IOException {
        Long earliest = changelogManager.earliestLongLivedChangelogId();
        Long latest = changelogManager.latestLongLivedChangelogId();
        if (earliest == null || latest == null) {
            return;
        }

        // it is possible that some snapshots have been expired
        List<Changelog> toBeCleaned = new ArrayList<>();
        long to = Math.max(earliest, retainedSnapshot.id() + 1);
        for (long i = latest; i >= to; i--) {
            toBeCleaned.add(changelogManager.changelog(i));
        }

        // modify the latest hint
        if (!toBeCleaned.isEmpty()) {
            if (to == earliest) {
                // all changelog has been cleaned, so we do not know the actual latest id
                // set to -1
                changelogManager.commitLongLivedChangelogLatestHint(-1);
            } else {
                changelogManager.commitLongLivedChangelogLatestHint(to - 1);
            }
        }

        // delete data files of changelog
        for (Changelog changelog : toBeCleaned) {
            fileIO.deleteQuietly(changelogManager.longLivedChangelogPath(changelog.id()));
        }
    }

    private void cleanTags(Snapshot retainedSnapshot) {
        SortedMap<Snapshot, List<String>> tags = tagManager.tags();
        if (tags.isEmpty()) {
            return;
        }

        List<Snapshot> taggedSnapshots = new ArrayList<>(tags.keySet());

        // delete tag files
        for (int i = taggedSnapshots.size() - 1; i >= 0; i--) {
            Snapshot tag = taggedSnapshots.get(i);
            if (tag.id() <= retainedSnapshot.id()) {
                break;
            }
            tags.get(tag).forEach(tagName -> fileIO.deleteQuietly(tagManager.tagPath(tagName)));
        }
    }
}

File: paimon-python/pypaimon/table/rollback_helper.py
"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from pypaimon.common.json_util import JSON


class RollbackHelper:
    """Helper class for table rollback including utils to clean snapshots.
    """

    def __init__(self, snapshot_manager, tag_manager, file_io):
        """Initialize RollbackHelper.

        Args:
            snapshot_manager: The SnapshotManager instance.
            tag_manager: The TagManager instance.
            file_io: The FileIO instance for file operations.
        """
        self._snapshot_manager = snapshot_manager
        self._tag_manager = tag_manager
        self._file_io = file_io

    def clean_larger_than(self, retained_snapshot):
        """Clean snapshots and tags whose id is larger than the retained snapshot.

        Updates the LATEST hint and removes snapshot files and tag files
        for snapshots newer than the retained one.

        Args:
            retained_snapshot: The snapshot to retain; everything newer is removed.
        """
        latest = self._snapshot_manager.get_latest_snapshot()
        if latest is None:
            return

        latest_id = latest.id
        retained_id = retained_snapshot.id

        # Update LATEST hint
        self._file_io.write_file(
            self._snapshot_manager.latest_file, str(retained_id), overwrite=True)

        # Delete snapshot files larger than retained
        for snapshot_id in range(retained_id + 1, latest_id + 1):
            snapshot_path = self._snapshot_manager.get_snapshot_path(snapshot_id)
            if self._file_io.exists(snapshot_path):
                self._file_io.delete(snapshot_path)

        # Clean tags whose snapshot id is larger than retained
        for tag_name in self._tag_manager.list_tags():
            tag = self._tag_manager.get(tag_name)
            if tag is not None:
                tag_snapshot = tag.trim_to_snapshot()
                if tag_snapshot.id > retained_id:
                    self._tag_manager.delete_tag(tag_name)

    def create_snapshot_file_if_needed(self, tagged_snapshot):
        """Create a snapshot file from a tag if the snapshot file doesn't exist.

        When rolling back to a tag, the snapshot file may have been expired.
        This method recreates it from the tag data and updates the earliest hint.

        Args:
            tagged_snapshot: The snapshot from the tag to potentially write.
        """
        snapshot_path = self._snapshot_manager.get_snapshot_path(tagged_snapshot.id)
        if not self._file_io.exists(snapshot_path):
            self._file_io.write_file(
                snapshot_path, JSON.to_json(tagged_snapshot), overwrite=False)
